### PR TITLE
Events

### PR DESCRIPTION
**Pull request by MCBEPluginModer**
Add PlayerBucketFillEvent,PlayerTameEntityEvent and PlayerTargetEntityEvent

### CODE REVIEW

**Comment by MCBEPluginModer** [comment]
Please move to LiteLoader:beta branch.

File: LiteLoader/Header/EventAPI.h [hunk 1]
```diff
@@ -700,9 +700,28 @@ class ConsoleOutputEvent : public EventTemplate<ConsoleOutputEvent>
...
 };
     
-class PlayerBedEnterEvent : : public EventTemplate<ConsoleOutputEvent> 
+class PlayerBedEnterEvent : public EventTemplate<ConsoleOutputEvent> 
 {
   Player* mPlayer;
   BlockInstance* mBlockInstance;
 };
+    
+class PlayerBucketFillEvent : public EventTemplate<ConsoleOutputEvent>
```
Comment (Jaciezyt) [inline LiteLoader/Header/EventAPI.h]:
```suggestion
class PlayerBucketFillEvent : public EventTemplate<PlayerBucketFillEvent>
```

File: LiteLoader/Header/EventAPI.h [hunk 2]
```diff
@@ -700,9 +700,28 @@ class ConsoleOutputEvent : public EventTemplate<ConsoleOutputEvent>
...
+{
+  Player* mPlayer;
+  ItemStack* mItemStack;
+  BlockInstance* mBlockInstance;
+};
+    
+class PlayerTameEntityEvent : public EventTemplate<ConsoleOutputEvent>
```
Comment (Jaciezyt) [inline LiteLoader/Header/EventAPI.h]:
```suggestion
class PlayerTameEntityEvent : public EventTemplate<PlayerTameEntityEvent>
```

File: LiteLoader/Header/EventAPI.h [hunk 3]
```diff
@@ -700,9 +700,28 @@ class ConsoleOutputEvent : public EventTemplate<ConsoleOutputEvent>
...
+{
+  Player* mPlayer;
+  Acctor* mActor;
+};
+
+class PlayerTargetEntityEvent : public EventTemplate<ConsoleOutputEvent>
```
Comment (Jaciezyt) [inline LiteLoader/Header/EventAPI.h]:
```suggestion
class PlayerTargetEntityEvent : public EventTemplate<PlayerTargetEntityEvent>
```

File: LiteLoader/Header/EventAPI.h [hunk 4]
```diff
@@ -700,9 +700,28 @@ class ConsoleOutputEvent : public EventTemplate<ConsoleOutputEvent>
     std::string mOutput;
 };
     
-class PlayerBedEnterEvent : : public EventTemplate<ConsoleOutputEvent> 
+class PlayerBedEnterEvent : public EventTemplate<ConsoleOutputEvent>
```
Comment (Jaciezyt) [inline LiteLoader/Header/EventAPI.h]:
```suggestion
class PlayerBedEnterEvent : public EventTemplate<PlayerBedEnterEvent> 
```

File: LiteLoader/Kernel/EventAPI.cpp [hunk 5]
```diff
@@ -122,6 +122,7 @@ DeclareEventListeners(ServerStartedEvent);
 DeclareEventListeners(ServerStoppedEvent);
 DeclareEventListeners(RegCmdEvent);
 DeclareEventListeners(PlayerBedEnterEvent);
+DeclareEventListeners(PlayerBucketFillEvent);
```
Comment (Jaciezyt) [inline LiteLoader/Kernel/EventAPI.cpp:125]:
```suggestion
DeclareEventListeners(PlayerBucketFillEvent);
// [Another two events]
```

File: LiteLoader/Header/EventAPI.h [hunk 6]
```diff
@@ -700,9 +700,32 @@ class ConsoleOutputEvent : public EventTemplate<ConsoleOutputEvent>
...
   Player* mPlayer;
+  ItemStack* mItemStack;
   BlockInstance* mBlockInstance;
 };
+    
+class PlayerTameEntityEvent : public EventTemplate<ConsoleOutputEvent>
+{
+public:
+  Player* mPlayer;
+  Acctor* mActor;
```
Comment (Jaciezyt) [inline LiteLoader/Header/EventAPI.h]:
```suggestion
  Actor* mActor;
```

File: LiteLoader/Kernel/EventAPI.cpp [hunk 7]
```diff
@@ -2009,3 +2010,47 @@ THook(int, "?startSleepInBed@Player@@UEAA?AW4BedSleepingResult@@AEBVBlockPos@@@Z
...
   return original(a1,a2);
 }
+
+THook(void, "?_takeLiquid@BucketItem@@AEBA_NAEAVItemStack@@AEAVActor@@AEBVBlockPos@@@Z", BucketItem* _this, ItemStack& s, Actor& a, const BlockPos& b)
+{
+    auto bl = Level::getBlockInstance(b,a.getDimensionId());;
+    IF_LISTENED(PlayerBucketFillEvent)
+    {
+        PlayerBucketFillEvent ev{};
+        ev.mPlayer = &a;
```
Comment (Jaciezyt) [inline LiteLoader/Kernel/EventAPI.cpp]:
```suggestion
        ev.mPlayer = (Player*)&a;
```

**Review by Jaciezyt — CHANGES_REQUESTED** [review]
TOO MANY BUGS
you should PR after the local build and avoid mistakes

**Comment by Jaciezyt** [comment]
Please clone your fork repo to your computer and use Visual Studio or CLion to modify your code and build the project.then `git commit&push`

File: LiteLoader/Kernel/EventAPI.cpp [hunk 8]
```diff
@@ -2009,3 +2012,47 @@ TInstanceHook(int, "?startSleepInBed@Player@@UEAA?AW4BedSleepingResult@@AEBVBloc
...
+{
+    auto bl = Level::getBlockInstance(b,a.getDimensionId());;
+    IF_LISTENED(PlayerBucketFillEvent)
+    {
+        PlayerBucketFillEvent ev{};
+        ev.mPlayer = (Player*)&a;
+        ev.mItemStack = (ItemStack*)&s;
+        ev.mBlockInstance = bl;
+        if (!ev.call())
+            return 0;
```
Comment (Jaciezyt) [inline LiteLoader/Kernel/EventAPI.cpp:2026]:
```suggestion
            return;
```

File: LiteLoader/Kernel/EventAPI.cpp [hunk 9]
```diff
@@ -2009,3 +2012,47 @@ TInstanceHook(int, "?startSleepInBed@Player@@UEAA?AW4BedSleepingResult@@AEBVBloc
...
 }
+
+THook(void, "?_takeLiquid@BucketItem@@AEBA_NAEAVItemStack@@AEAVActor@@AEBVBlockPos@@@Z", BucketItem* _this, ItemStack& s, Actor& a, const BlockPos& b)
+{
+    auto bl = Level::getBlockInstance(b,a.getDimensionId());;
+    IF_LISTENED(PlayerBucketFillEvent)
+    {
+        PlayerBucketFillEvent ev{};
+        ev.mPlayer = (Player*)&a;
+        ev.mItemStack = (ItemStack*)&s;
```
Comment (Jaciezyt) [inline LiteLoader/Kernel/EventAPI.cpp:2023]:
```suggestion
        ev.mItemStack = &s;
```

File: LiteLoader/Kernel/EventAPI.cpp [hunk 10]
```diff
@@ -2009,3 +2012,47 @@ TInstanceHook(int, "?startSleepInBed@Player@@UEAA?AW4BedSleepingResult@@AEBVBloc
...
+
+THook(void, "?_becomeTame@TameableComponent@@AEAAXAEAVActor@@@Z", TameableComponent* _this, Actor& act)
+{
+    IF_LISTENED(PlayerTameEntityEvent)
+    {
+        PlayerTameEntityEvent ev{};
+        ev.mPlayer = act.getPlayerOwner();
+        ev.mActor = act;
+        if (!ev.call())
+            return 0;
```
Comment (Jaciezyt) [inline LiteLoader/Kernel/EventAPI.cpp:2040]:
```suggestion
            return;
```

File: LiteLoader/Kernel/EventAPI.cpp [hunk 11]
```diff
@@ -2009,3 +2012,47 @@ TInstanceHook(int, "?startSleepInBed@Player@@UEAA?AW4BedSleepingResult@@AEBVBloc
...
+    return original(_this,s,a,b);
+}
+
+THook(void, "?_becomeTame@TameableComponent@@AEAAXAEAVActor@@@Z", TameableComponent* _this, Actor& act)
+{
+    IF_LISTENED(PlayerTameEntityEvent)
+    {
+        PlayerTameEntityEvent ev{};
+        ev.mPlayer = act.getPlayerOwner();
+        ev.mActor = act;
```
Comment (Jaciezyt) [inline LiteLoader/Kernel/EventAPI.cpp:2038]:
```suggestion
        ev.mActor = &act;
```

File: LiteLoader/Kernel/EventAPI.cpp [hunk 12]
```diff
@@ -2009,3 +2012,47 @@ TInstanceHook(int, "?startSleepInBed@Player@@UEAA?AW4BedSleepingResult@@AEBVBloc
...
+
+THook(bool, "?isValidTarget@ServerPlayer@@UEBA_NPEAVActor@@@Z", ServerPlayer* _this, Actor* mob)
+{
+    IF_LISTENED(PlayerTargetEntityEvent)
+    {
+        PlayerTameEntityEvent ev{};
+        ev.mPlayer = (Player*)_this;
+        ev.mActor = mob;
+        if (!ev.call())
+            return 0;
```
Comment (Jaciezyt) [inline LiteLoader/Kernel/EventAPI.cpp:2054]: Use false is better
```suggestion
            return false;
```

**Review by Jaciezyt — CHANGES_REQUESTED** [review]

**Comment by KawaiiNahida** [comment]
Build and test what you have wrote

**Comment by Jaciezyt** [comment]
You can start the actions in your repo

**Comment by ghost** [comment]
Zhenya...

**Pull request closed without merging**